### PR TITLE
[AIR][CUJ] Make distributing training benchmark at silver tier

### LINKED ISSUE #26630: [AIR][Train] TorchTrainer very slow data ingest with multiple GPU workers
### What happened + What you expected to happen

Setup 8 g3.8xlarge hosts with autoscaler with 20GB open dataset. We're scheduling all GPU workers but after data reading, 20mins+ passed we still didn't see significant GPU utils. In the meantime same experiment one single GPU finished in another cluster.

Screen Shot 2022-07-16 at 12 15 23 PM

Screen Shot 2022-07-16 at 12 15 27 PM

New experiment with 4 of g3.16xlarge instances 

```
======== Object references status: 2022-07-16 16:07:25.804096 ========
--- Aggregate object store stats across all nodes ---
Plasma memory usage 97694 MiB, 2033 objects, 16.57% full, 16.45% needed
Objects consumed by Ray tasks: 2710 MiB.
```

After initial peak load of network and CPU, then seems like trainers are just .... chilling .... ???

Screen Shot 2022-07-16 at 4 07 00 PM

### Versions / Dependencies

on master

### Reproduction script

```
import click
import time
import json
import os
import numpy as np
import pandas as pd
from io import BytesIO
from typing import List

from PIL import Image
from torchvision import transforms
from torchvision.models import resnet18
import torch
import torch.nn as nn
import torch.optim as optim

import ray
from ray.air.util.tensor_extensions.pandas import TensorArray
from ray.train.torch import to_air_checkpoint
from ray.data.preprocessors import BatchMapper
from ray import train
from ray.air import session
from ray.train.torch import TorchTrainer


# TODO(jiaodong): Remove this once ImageFolder #24641 merges
def convert_to_pandas(byte_item_list: List[bytes]) -> pd.DataFrame:
    """
    Convert input bytes into pandas DataFrame with image column and value of
    TensorArray to prevent serializing ndarray image data.
    """
    images = [
        Image.open(BytesIO(byte_item)).convert("RGB") for byte_item in byte_item_list
    ]
    images = [np.asarray(image) for image in images]
    # Dummy label since we're only testing training throughput
    labels = [1 for _ in range(len(images))]

    return pd.DataFrame({"image": TensorArray(images), "label": labels})


def preprocess(df: pd.DataFrame) -> pd.DataFrame:
    """
    User Pytorch code to transform user image. Note we still use pandas as
    intermediate format to hold images as shorthand of python dictionary.
    """
    preprocess = transforms.Compose(
        [
            transforms.ToTensor(),
            transforms.Resize(256),
            transforms.CenterCrop(224),
            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
        ]
    )
    df["image"] = df["image"].map(preprocess)
    df["image"] = df["image"].map(lambda x: x.numpy())
    df["image"] = TensorArray(df["image"])
    return df


def train_loop_per_worker(config):
    raw_model = resnet18(pretrained=True)
    model = train.torch.prepare_model(raw_model)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)

    train_dataset_shard = session.get_dataset_shard("train")

    for epoch in range(config["num_epochs"]):
        running_loss = 0.0
        for i, data in enumerate(
            train_dataset_shard.iter_batches(
                batch_size=config["batch_size"], batch_format="numpy"
            )
        ):
            # get the inputs; data is a list of [inputs, labels]
            inputs = torch.as_tensor(data["image"], dtype=torch.float32).to(
                device="cuda"
            )
            labels = torch.as_tensor(data["label"], dtype=torch.int64).to(device="cuda")
            # zero the parameter gradients
            optimizer.zero_grad()

            # forward + backward + optimize
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            loss.backward()
            optimizer.step()

            # print statistics
            running_loss += loss.item()
            if i % 2000 == 1999:  # print every 2000 mini-batches
                print(f"[{epoch + 1}, {i + 1:5d}] loss: {running_loss / 2000:.3f}")
                running_loss = 0.0

        session.report(
            dict(running_loss=running_loss),
            checkpoint=to_air_checkpoint(model),
        )


@click.command(help="Run Batch prediction on Pytorch ResNet models.")
@click.option("--data-size-gb", type=int, default=1)
@click.option("--num-epochs", type=int, default=2)
def main(data_size_gb: int, num_epochs=2):
    data_url = f"s3://air-example-data-2/{data_size_gb}G-image-data-synthetic-raw"
    print(
        "Running Pytorch image model training with "
        f"{data_size_gb}GB data from {data_url}"
    )
    print(f"Training for {num_epochs} epochs.")
    start = time.time()
    dataset = ray.data.read_binary_files(paths=data_url)
    # TODO(jiaodong): Remove this once ImageFolder #24641 merges
    dataset = dataset.map_batches(convert_to_pandas)

    preprocessor = BatchMapper(preprocess)

    trainer = TorchTrainer(
        train_loop_per_worker=train_loop_per_worker,
        train_loop_config={"batch_size": 64, "num_epochs": num_epochs},
        datasets={"train": dataset},
        preprocessor=preprocessor,
        scaling_config={"num_workers": int(ray.available_resources()["GPU"]), "use_gpu": True},
    )
    trainer.fit()

    total_time_s = round(time.time() - start, 2)

    # For structured output integration with internal tooling
    results = {"data_size_gb": data_size_gb, "num_epochs": num_epochs}
    results["perf_metrics"] = [
        {
            "perf_metric_name": "total_time_s",
            "perf_metric_value": total_time_s,
            "perf_metric_type": "LATENCY",
        },
        {
            "perf_metric_name": "throughout_MB_s",
            "perf_metric_value": round(
                num_epochs * data_size_gb * 1024 / total_time_s, 2
            ),
            "perf_metric_type": "THROUGHPUT",
        },
    ]

    test_output_json = os.environ.get("TEST_OUTPUT_JSON", "/tmp/release_test_out.json")
    with open(test_output_json, "wt") as f:
        json.dump(results, f)

    print(results)


if __name__ == "__main__":
    main()

```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

In previous setup, our script hangs due to missing `NCCL_SOCKET_IFNAME ` flag. Adding this back unblocks cross-host GPU training.

System metrics much better now:

Screen Shot 2022-07-16 at 7 30 17 PM

Also added compute template with 16 GPUs, added silver tier test for both prediction.

## Related issue number

Closes #26630

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py
Comment (amogkam) [inline release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py:95]: Should we update the nightly test to run with the silver configuration?

Comment (jiaodong) [reply release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py:95]: hmm ideally yes but i wasn't sure what our config yaml should be for VPC / secure group to ensure cross host NCCL. My workspaces had a PlacementGroup with secure group setup by default that took care of it.

Comment (amogkam) [reply release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py:95]: hmm is specifying the runtime env not sufficient?

File: doc/source/ray-air/benchmarks.rst
Comment (richardliaw) [inline doc/source/ray-air/benchmarks.rst:148]: also they should also be in the same placement group?

File: doc/source/ray-air/benchmarks.rst
Comment (richardliaw) [inline doc/source/ray-air/benchmarks.rst:165]: nodes

**Review by richardliaw — COMMENTED** [review]
will leave it to @amogkam to approve

File: doc/source/ray-air/benchmarks.rst
Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:148]: yes that's what i had in my workspaces too. Do you know any train's existing tests that had the right YAML setup for this ?

Comment (amogkam) [reply doc/source/ray-air/benchmarks.rst:148]: Yes this tests is run with multi-host GPU training: https://github.com/ray-project/ray/blob/master/release/release_tests.yaml#L925-L947

File: release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py
Comment (jiaodong) [reply release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py:95]: you're right i only had it in my workspaces, looking for full YAML config for nightly now ..

File: doc/source/ray-air/benchmarks.rst
Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:148]: hmm that seems to point to a CPU compute template. Let me create a public PlacementGroup and secure group on CI

Comment (amogkam) [reply doc/source/ray-air/benchmarks.rst:148]: This is the compute template that it's using https://github.com/ray-project/ray/blob/master/release/ml_user_tests/train/compute_tpl.yaml

Uses 3 g3.8xlarge isntances

Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:148]: right, i didn't find any config regarding their secure group or placement group tho, am i missing something here or it's not needed for our previous multi-host gpu tests

Comment (amogkam) [reply doc/source/ray-air/benchmarks.rst:148]: Ah that's right- what I'm saying is I don't think it's needed if we set the runtime env

Comment (jiaodong) [reply doc/source/ray-air/benchmarks.rst:148]: i see what you mean now, makes sense and let me update the PR :)

**Review by amogkam — APPROVED** [review]
Thanks @jiaodong! Lgtm once the tests pass

**Review by ericl — APPROVED** [review]

**Comment by richardliaw** [comment]
release test fails for some odd reason unrelated to this pr... let's merge then see what happens

**Pull request merged**
